decision-records/2025-03-20--node-exporter-otelcol-adr.md

```diff
@@ -0,0 +1,170 @@
+
+If the idea is so simple, why wouldn't we implement it? Well, the answer is that from a charm perspective is not that simple. In fact in `grafana-agent` charm it is not implemented, and because of this we have [some weird situations.](https://discourse.charmhub.io/t/one-grafana-agent-charm-to-rule-them-all/16014).
+
+### Some considerations to be taken into account when implementing this solution
+
+
+Everytime `cos-collector` subordinate charm is related to a principal charm, `cos-collector` charm must:
+
+* Verify whether `otelcol` and `node-exporter` snaps are already installed or not to avoid trying to install them again.
+* Merge the `otelcol` configuration resulting from the established relationship with any previously existing configuration.
+
+
+When a relation between `cos-collector` and a principal charm is removed, the `cos-collector` charm must:
+
+* Remove from the `otelcol` config file the configuration resulting from the departing relation.
+* Uninstall `otelcol` and `node-exporter` snaps only if there is no other `cos-collector` charm deployed in the same host.
+
+
+#### Questions and doubts about this approach
+
+
+* When we think of a charm, we expect it to manage the lifecycle of an application, from installation, configuration, scaling to relationships. However, with this approach, multiple instances of `cos-collector` charm will handle a single binary and a single configuration file.
+
+
+[![](https://mermaid.ink/img/pako:eNqlU01zgjAQ_SvMnsERRIEcPPTj1l7antr0ECFCppAwIUy16n_vAoJjxU5nmtOy-97btxuyg1glHAisc_UZZ0wb6-GJyqpepZqVmZWpylBp4UmE5rERSlovN1R2uQEWq8qJVZ4jQmmn0Sk6xCWzz7Ygp-l-yg16EtMO35RKG66dSrLyBGrOWd19o3CWsFZCMr2l8P4LC1vLtUgvuF261qxxfCbBZTJiVRme4-gjJo8V7HCMrvnqJQZHPf6PXk6rtBxn-UPuKmxsGx24Fe_CkYtFsrPcU8ANKJ3gRIZT2FulFjIWJcs7lPtPvkcl2gAbCq4LJhL8RXeNJAWT8QIZBMOE6Q8KVB4QV5cJKt0nAhsBWbO84jaw2qjnrYyBGF3zHnQnGN5dMaBKJoHsYANkPptEzVm44Xw29YKZb8MWCH5NFv504c0PNnwphdTpJAz8BhpEbhSFYRDZwNvmj92Dat9Vq_3aEjoDWtVpNjROdTPYsYLjcn2rammAeH54-AalNSo9?type=png)](https://mermaid.live/edit#pako:eNqlU01zgjAQ_SvMnsERRIEcPPTj1l7antr0ECFCppAwIUy16n_vAoJjxU5nmtOy-97btxuyg1glHAisc_UZZ0wb6-GJyqpepZqVmZWpylBp4UmE5rERSlovN1R2uQEWq8qJVZ4jQmmn0Sk6xCWzz7Ygp-l-yg16EtMO35RKG66dSrLyBGrOWd19o3CWsFZCMr2l8P4LC1vLtUgvuF261qxxfCbBZTJiVRme4-gjJo8V7HCMrvnqJQZHPf6PXk6rtBxn-UPuKmxsGx24Fe_CkYtFsrPcU8ANKJ3gRIZT2FulFjIWJcs7lPtPvkcl2gAbCq4LJhL8RXeNJAWT8QIZBMOE6Q8KVB4QV5cJKt0nAhsBWbO84jaw2qjnrYyBGF3zHnQnGN5dMaBKJoHsYANkPptEzVm44Xw29YKZb8MWCH5NFv504c0PNnwphdTpJAz8BhpEbhSFYRDZwNvmj92Dat9Vq_3aEjoDWtVpNjROdTPYsYLjcn2rammAeH54-AalNSo9)
+
+## Decision
+
+TBC.
```

Comment:
Do you have any recommendation, based on the content of the ADR?

Reply:
Yes, I have a recommendation.
I didn't want to write it down **right now** to avoid influencing the recommendations that the rest of the team might make.